waku/v2/api/publish/message_check.go

```diff
@@ -0,0 +1,198 @@
+package publish
+
+import (
+	"bytes"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/pb"
+	"github.com/waku-org/go-waku/waku/v2/protocol/store"
+	"github.com/waku-org/go-waku/waku/v2/timesource"
+	"go.uber.org/zap"
+)
+
+const maxHashQueryLength = 100
+const hashQueryInterval = 3 * time.Second
+const messageSentPeriod = 3    // in seconds
+const messageExpiredPerid = 10 // in seconds
+
+type MessageSentCheck struct {
+	messageIDs         map[string]map[common.Hash]uint32
+	messageIDsMu       sync.RWMutex
+	storePeerID        peer.ID
+	MessageStoredChan  chan common.Hash
+	MessageExpiredChan chan common.Hash
+	ctx                context.Context
+	store              *store.WakuStore
+	timesource         timesource.Timesource
+	logger             *zap.Logger
+}
+
+func NewMessageSentCheck(ctx context.Context, store *store.WakuStore, timesource timesource.Timesource, logger *zap.Logger) *MessageSentCheck {
+	return &MessageSentCheck{
+		messageIDs:         make(map[string]map[common.Hash]uint32),
+		messageIDsMu:       sync.RWMutex{},
+		MessageStoredChan:  make(chan common.Hash, 1000),
+		MessageExpiredChan: make(chan common.Hash, 1000),
+		ctx:                ctx,
+		store:              store,
+		timesource:         timesource,
+		logger:             logger,
+	}
+}
+
+func (m *MessageSentCheck) Add(topic string, messageID common.Hash, sentTime uint32) {
+	m.messageIDsMu.Lock()
+	defer m.messageIDsMu.Unlock()
+
+	if _, ok := m.messageIDs[topic]; !ok {
+		m.messageIDs[topic] = make(map[common.Hash]uint32)
+	}
+	m.messageIDs[topic][messageID] = sentTime
+}
+
+func (m *MessageSentCheck) DeleteByMessageIDs(messageIDs []common.Hash) {
```

Comment:
When will this be used?

Reply:
When MVDS confirmation is received, it will call this method to removing the messageID.